cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -0,0 +1,136 @@
+import com.google.inject.name.Named;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.client.HBaseAdmin;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Handles upgrade for System and User Datasets
+ */
+public class DatasetUpgrader extends AbstractUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetUpgrader.class);
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final LocationFactory locationFactory;
+  private final QueueAdmin queueAdmin;
+  private final HBaseTableUtil hBaseTableUtil;
+  private final DatasetFramework namespacedFramework;
+
+  @Inject
+  private DatasetUpgrader(CConfiguration cConf, Configuration hConf, LocationFactory locationFactory,
+                          QueueAdmin queueAdmin, HBaseTableUtil hBaseTableUtil,
+                          @Named("namespacedDSFramework") DatasetFramework namespacedFramework) {
+
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.locationFactory = locationFactory;
+    this.queueAdmin = queueAdmin;
+    this.hBaseTableUtil = hBaseTableUtil;
+    this.namespacedFramework = namespacedFramework;
+  }
+
+  @Override
+  public void upgrade() throws Exception {
+    // Upgrade system dataset
+    upgradeSystemDatasets(namespacedFramework);
+
+    // Upgrade all user hbase tables
+    upgradeUserTables();
+
+    // Upgrade all queue and stream tables.
+    queueAdmin.upgrade();
+  }
+
+  private void upgradeSystemDatasets(DatasetFramework framework) throws Exception {
+
+    // Upgrade all datasets in system namespace
+    Id.Namespace systemNamespace = Constants.SYSTEM_NAMESPACE_ID;
+    for (DatasetSpecification spec : framework.getInstances(systemNamespace)) {
+      LOG.info("Upgrading dataset: {}, spec: {}", spec.getName(), spec.toString());
+      DatasetAdmin admin = framework.getAdmin(Id.DatasetInstance.from(systemNamespace, spec.getName()), null);
+      // we know admin is not null, since we are looping over existing datasets
+      admin.upgrade();
+      LOG.info("Upgraded dataset: {}", spec.getName());
+    }
+  }
+
+  private void upgradeUserTables() throws Exception {
+    DefaultDatasetNamespace namespace = new DefaultDatasetNamespace(cConf);
+    HBaseAdmin hAdmin = new HBaseAdmin(hConf);
+
+
+    for (HTableDescriptor desc : hAdmin.listTables()) {
+      String tableName = desc.getNameAsString();
+      TableId tableId = TableId.from(tableName);
+      Id.DatasetInstance datasetInstanceId = Id.DatasetInstance.from(tableId.getNamespace(), tableId.getTableName());
+      // todo: it works now, but we will want to change it if namespacing of datasets in HBase is more than +prefix
+      if (namespace.fromNamespaced(datasetInstanceId) != null) {
+        LOG.info("Upgrading hbase table: {}, desc: {}", tableName, desc);
```

Comment:
did you test this ? , in https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/DefaultDatasetNamespace.java#L57
we use System namespace, while we expect to upgrade user tables here.